### PR TITLE
Cache RoPE freqs on device to avoid repeated CPU-GPU copy in QwenImage

### PR DESCRIPTION
**Pull request by akshan-main**
## What does this PR do?

Part of #13401

`QwenEmbedRope.forward()` copies `pos_freqs` and `neg_freqs` from CPU to GPU via `.to(device)` on every transformer forward call. These tensors are fixed at init and never change, so the repeated transfer triggers an unnecessary `cudaStreamSynchronize` (~76ms each).

Added `_get_device_freqs()` that caches the GPU copy on first call. Applied to both `QwenEmbedRope` and `QwenEmbedLayer3DRope`.

(`register_buffer` can't be used here because it drops the imaginary part of complex tensors)

### Profiling (A100 80GB, eager, 2 steps, 1024x1024)

```
                                     BEFORE        AFTER
------------------------------ ------------ ------------
Big syncs (>50ms)                         3            0
Big sync total (ms)                   228.7          0.0
Big syncs before: [76.6, 76.4, 75.7]
Big syncs after:  []
```

**Before** (76ms cudaStreamSynchronize inside transformer_forward):

before_sync

**After** (no sync gap):

after_sync

Profiled with the tooling from #13356. [Reproduction notebook](https://colab.research.google.com/gist/akshan-main/bd8a031fbe3f601c2d2194c643ea0ea4/qwenimage_rope_freq_cache_profiling.ipynb).

Part of #13401

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case. #13401
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@sayakpaul @dg845

### CODE REVIEW

**Comment by akshan-main** [comment]
The profiling was done with 2 steps, but this sync happens every transformer forward call, so at 20 inference steps, this eliminates ~1.5s of CPU-GPU sync overhead per run. Under torch.compile the impact is larger since GPU queues are deeper(each sync stalls longer) (80ms vs 76ms in eager).

**Comment by akshan-main** [comment]
oh and this fix applies to all QwenImage variants (Edit, EditPlus, Layered) since they share the same transformer

**Comment by sayakpaul** [comment]
@akshan-main thanks for this! In the second plot, could you tell which one of the blocks the reported duration belongs to?

**Review by sayakpaul — APPROVED** [review]
Seems like a clean fix to me. But I will let @dg845 make the final merge.

**Comment by akshan-main** [comment]
the selected slice in after image is the transformer_forward user_annotation itself (~439ms), wrapping the full QwenImageTransformer2DModel.forward. 

I am highlighting a specific sub-block showing where the 76ms cudaStreamSynchronize used to sit (in the before screenshot) is gone.

**Comment by akshan-main** [comment]
~439ms is for entire transformer_forward block

**Comment by akshan-main** [comment]
Friendly ping @dg845, hey! seeking your review/ interpretation

File: src/diffusers/models/transformers/transformer_qwenimage.py [hunk 1]
```diff
@@ -233,6 +234,12 @@ def rope_params(self, index, dim, theta=10000):
         freqs = torch.polar(torch.ones_like(freqs), freqs)
         return freqs
 
+    def _get_device_freqs(self, device: torch.device) -> tuple[torch.Tensor, torch.Tensor]:
+        """Return pos_freqs and neg_freqs on the given device, caching the transfer."""
+        if device not in self._device_freq_cache:
+            self._device_freq_cache[device] = (self.pos_freqs.to(device), self.neg_freqs.to(device))
+        return self._device_freq_cache[device]
```
Comment (dg845) [inline src/diffusers/models/transformers/transformer_qwenimage.py]:
```suggestion
    @lru_cache_unless_export(maxsize=None)
    def _get_device_freqs(self, device: torch.device) -> tuple[torch.Tensor, torch.Tensor]:
        """Return pos_freqs and neg_freqs on the given device."""
        return self.pos_freqs.to(device), self.neg_freqs.to(device)
```
I think this might be slightly cleaner since `lru_cache_unless_export` should handle the cases where we're compiling or exporting the model correctly.

File: src/diffusers/models/transformers/transformer_qwenimage.py [hunk 2]
```diff
@@ -222,6 +222,7 @@ def __init__(self, theta: int, axes_dim: list[int], scale_rope=False):
 
         # DO NOT USING REGISTER BUFFER HERE, IT WILL CAUSE COMPLEX NUMBERS LOSE ITS IMAGINARY PART
         self.scale_rope = scale_rope
+        self._device_freq_cache: dict[torch.device, tuple[torch.Tensor, torch.Tensor]] = {}
```
Comment (dg845) [inline src/diffusers/models/transformers/transformer_qwenimage.py]:
```suggestion
```
Follow-up change to https://github.com/huggingface/diffusers/pull/13406#discussion_r3055842846.

**Comment by dg845** [comment]
@bot /style

**Review by dg845 — COMMENTED** [review]
Thanks for the PR! Left one suggestion about using `lru_cache_unless_export` instead of caching manually.

**Comment by akshan-main** [comment]
@dg845 done! switched both QwenEmbedRope and QwenEmbedLayer3DRope to lru_cache_unless_export
ci testing shouldn't be a issue too now

**Review by dg845 — APPROVED** [review]
Thanks!

**Comment by dg845** [comment]
Hi @akshan-main, have you also profiled the QwenImage pipeline when using `torch.compile`? Are the CPU to GPU syncs also eliminated by this fix in that case?

**Comment by akshan-main** [comment]
profiling compile before/after now, will update with numbers

**Comment by akshan-main** [comment]
@dg845  profiled compile before/after. torch.compile() already eliminates the big syncs on its own (0 big syncs in both before and after). The fix specifically targets eager mode

**Comment by akshan-main** [comment]
[Reproduction notebook for compile](https://colab.research.google.com/gist/akshan-main/a698d0a2feaff95d70521edd75c35df2/qwenimage_rope_freq_cache_profiling.ipynb)

**Comment by akshan-main** [comment]
@sayakpaul @dg845 don't think the failures are related to my PR

**Comment by sayakpaul** [comment]
I will let @dg845 take care of the final merging. I am looking into the failing tests (unrelated to your PR).

I also got this script to compare QwenImage outputs on this branch and `main` branch and they pass:

```py
"""Compare QwenImagePipeline outputs between current branch and main."""

import subprocess
import sys

import torch
from diffusers import DiffusionPipeline


REPO_ID = "Qwen/Qwen-Image"
PROMPT = "A cat holding a sign that says hello world"
NUM_INFERENCE_STEPS = 2
HEIGHT = 256
WIDTH = 256


def get_output(pipe):
    output = pipe(
        PROMPT,
        num_inference_steps=NUM_INFERENCE_STEPS,
        height=HEIGHT,
        width=WIDTH,
        generator=torch.manual_seed(0),
        output_type="np",
    ).images[0]
    return output


def main():
    current_branch = (
        subprocess.check_output(["git", "rev-parse", "--abbrev-ref", "HEAD"]).decode().strip()
    )
    print(f"Current branch: {current_branch}")

    # --- Current branch ---
    print("Loading pipeline on current branch...")
    pipe = DiffusionPipeline.from_pretrained(REPO_ID, torch_dtype=torch.bfloat16).to("cuda")
    print("Computing output on current branch...")
    output_current = get_output(pipe)

    del pipe
    torch.cuda.empty_cache()

    # --- main branch ---
    print("Checking out main branch...")
    subprocess.check_call(["git", "checkout", "main"])

    # Reload diffusers from main
    subprocess.check_call([sys.executable, "-m", "pip", "install", "-e", ".", "--quiet"])

    print("Loading pipeline on main branch...")
    pipe = DiffusionPipeline.from_pretrained(REPO_ID, torch_dtype=torch.bfloat16).to("cuda")
    print("Computing output on main branch...")
    output_main = get_output(pipe)

    del pipe
    torch.cuda.empty_cache()

    # --- Restore original branch ---
    print(f"Restoring branch: {current_branch}")
    subprocess.check_call(["git", "checkout", current_branch])

    # --- Compare ---
    max_diff = abs(output_current - output_main).max()
    mean_diff = abs(output_current - output_main).mean()

    print(f"\nMax absolute difference:  {max_diff}")
    print(f"Mean absolute difference: {mean_diff}")

    if max_diff < 1e-3:
        print("PASSED: Outputs match.")
    else:
        print("FAILED: Outputs differ significantly.")


if __name__ == "__main__":
    main()

```

**Comment by akshan-main** [comment]
@sayakpaul outputs should match since the fix only changes how the freqs are cached, not the computation itself

**Comment by akshan-main** [comment]
friendly ping @dg845 same tests are failing

**Review by dg845 — APPROVED** [review]
Thanks!

**Comment by dg845** [comment]
Merging as the CI failures should be unrelated.

**Comment by akshan-main** [comment]
thanks @sayakpaul @dg845! for the opportunity to contribute

**Comment by akshan-main** [comment]
Hey @sayakpaul @dg845, I am very much interested in this theme of performance engineering fixes. I was wondering what pipeline you would want me to look into next so that I can contribute promptly.

**Comment by sayakpaul** [comment]
It could be nice to continue on the Qwen series? WDYT?

**Comment by akshan-main** [comment]
@sayakpaul I can do that. I will work on QwenImageEdit next

**Comment by akshan-main** [comment]
@sayakpaul @dg845 

just done profiling QwenImageEdit (eager, 2 steps). after the RoPE fix, remaining syncs are ~150ms from masked boolean indexing in _extract_masked_hidden (~80ms) and torch.tensor(config.latents_mean).to(device) in _encode_vae_image (~66ms). These are not fixable without restructuring the masking approach. QwenImageEdit looks clean otherwise.

Notebook link where I verified:
https://colab.research.google.com/gist/akshan-main/e7f85bb6996908e9c62ee1797f287f26/qwenimageedit_profile.ipynb
